SelfAndPeerEvaluations/utils.py

```diff
@@ -0,0 +1,439 @@
+            self.__give_solo_submission_scores(json_file, row)
+        else:
+            self.__give_quiz_scores(json_file, row)
+        self.detailed_scores.append(row)
+
+    def __stringify_subjects(self, subjects: list) -> str:
+        result = "("
+        cur_line = ""
+        for subject in subjects:
+            if subject == subjects[-1]:
+                if len(cur_line + subject) >= 45:
+                    result += cur_line + "\n" + subject + ")"
+                else:
+                    result += cur_line + subject + ")"
+            else:
+                if len(cur_line + subject) >= 45:
+                    result += cur_line + "\n"
+                    cur_line = subject + ", "
+                else:
+                    cur_line += subject + ", "
+        return result
+
+    def __make_detailed_header(self, json_file: dict) -> list:
+        header = ["Categories"]
+        for qualitative_subject in json_file["info"]["qualitative_subjects"]:
+            header.append(qualitative_subject)
+        average_qualitative_difference_str = f"Average Qualitative Difference\n{self.__stringify_subjects(json_file['info']['qualitative_subjects'])}"
+        header.append(average_qualitative_difference_str)
+        # since there's only one quantitative_overall_subject (contribution)
+        # we won't have an "Average Quantitative Overall Difference" column
+        header.append("Project Contribution")
+        return header
+
+
+def find_partner_eval_ag(
+    course: canvasapi.course.Course,
+) -> canvasapi.assignment.AssignmentGroup:
+    # returns AssignmentGroup named "Partner Evaluations"
+    # ends program if no AssignmentGroup is found
+    partner_eval_ag = None
+    for ag in course.get_assignment_groups():
+        if ag.name == "Partner Evalulations":
+            partner_eval_ag = ag
+
+    if partner_eval_ag is None:
+        print('"Partner Evaluations" category does not exist!')
+        print("Cannot grade, so ending program now")
+        exit(1)
+
+    return course.get_assignment_group(partner_eval_ag.id, include=["assignments"])
+
+
+def make_student_id_map(course: canvasapi.course.Course) -> dict:
+    # initially creates a dict where keys are student names and value is list [SID, canvas ID]
+    # returns a dict where keys are student names and value is canvas ID (2nd element)
+    """
+    Creates a mapping of unique student sortable names to canvas ID's
+    If student sortable names are duplicate, then last 4 or 5 digits are appended to student sortable name
+    :returns: dict where key is unique sortable name and value is canvas ID
+    """
+    raw_students = defaultdict(list)
+    dup_list = set()
+    for raw_student in course.get_users(sort="username", enrollment_type=["student"]):
+        if (
+            raw_student.sis_user_id is None
+        ):  # FIXME: delete for real class; not all sandbox student have SID
+            raw_student.sis_user_id = random.randint(0, 100)
```

Comment:
This seems wrong. If they have to have an int for an id, make it something impossible like a weird negative number like -68945